calico-vpp-agent/watchers/nodes_watcher.go

```diff
@@ -279,55 +214,81 @@ func (w *NodeWatcher) configureRemoteNodeSnat(node *common.NodeState, isAdd bool
 	}
 }
 
-func (w *NodeWatcher) onNodeDeleted(old *common.NodeState) error {
+func (w *NodeWatcher) onNodeDeleted(old *oldv3.Node) error {
 	common.SendEvent(common.CalicoVppEvent{
 		Type: common.PeerNodeStateChanged,
-		Old:  &old.Node,
+		Old:  old,
 	})
+	if old.Name == config.NodeName {
+		// restart if our BGP config changed
+		return NodeWatcherRestartError{}
+	}
+
 	w.configureRemoteNodeSnat(old, false /* isAdd */)
 	return nil
 }
 
-func (w *NodeWatcher) onNodeUpdated(old *common.NodeState, node *common.NodeState) (err error) {
+type NodeWatcherRestartError struct{}
+
+func (e NodeWatcherRestartError) Error() string {
+	return "node configuration changed, restarting"
+}
+
+func (w *NodeWatcher) onNodeUpdated(old *oldv3.Node, node *oldv3.Node) (err error) {
 	w.log.Debugf("node comparison: old:%+v new:%+v", old.Spec.BGP, node.Spec.BGP)
 
-	newV4IP, newV6IP := common.GetNodeSpecAddresses(&node.Node)
-	oldV4IP, oldV6IP := common.GetNodeSpecAddresses(&old.Node)
+	newV4IP, newV6IP := common.GetNodeSpecAddresses(node)
+	oldV4IP, oldV6IP := common.GetNodeSpecAddresses(old)
 
 	// This is used by the routing server to process Wireguard key updates
 	// As a result we only send an event when a node is updated, not when it is added or deleted
 	common.SendEvent(common.CalicoVppEvent{
 		Type: common.PeerNodeStateChanged,
-		Old:  &old.Node,
-		New:  &node.Node,
+		Old:  old,
+		New:  node,
 	})
 
-	if common.GetStringChangeType(oldV4IP, newV4IP) > common.ChangeSame ||
-		common.GetStringChangeType(oldV6IP, newV6IP) > common.ChangeSame {
+	change := common.MaxChangeType(common.GetStringChangeType(oldV4IP, newV4IP), common.GetStringChangeType(oldV6IP, newV6IP))
```

Comment:
What does the change order mean? I think we should make this easier to understand.

Reply:
Agreed, switched to a flag based logic, should be more straightforward than the ordering

Reply:
Way easier to understand, thanks!